src/background/host.ts

```diff
@@ -1,26 +1,87 @@
+import Debug from "debug";
 import Peer from "peerjs";
-import { Message, MessageType } from "../Message";
+import { Friend, LocalOutMessage, LocalPollMessage, MessageType, RemoteMessage } from "../Message";
+
+export interface FriendConnected extends Friend {
+    conn: Peer.DataConnection;
+}
 
 export class Host {
+    #log = Debug("peer:disconnected");
     #port: chrome.runtime.Port;
     #peer = new Peer();
-    #friends = new Set<Peer.DataConnection>();
 
-    public isReady = false;
-    public id: string | null = null;
+    #friends = new Map<string, FriendConnected>();
+
+    public personalData: Friend | null = null;
     // eslint-disable-next-line @typescript-eslint/no-empty-function
     public onChangeFriends: (count: number, delta: number) => void = () => {};
 
-    #handleConnect = (conn: Peer.DataConnection): void => {
-        this.#friends.add(conn);
-        this.onChangeFriends(this.#friends.size, 1);
+    #ensureConnections = (friends: Friend[], conn: Peer.DataConnection): void => {
+        if (!this.personalData) {
+            return;
+        }
+
+        for (const member of friends) {
+            const friend = this.#friends.get(member.id);
+
+            if (member.id === this.personalData.id) {
+                // do nothing regarding our own connection
+            } else if (friend) {
+                // update member data without touching `friend.conn`
+                this.#friends.set(member.id, {
+                    ...member,
+                    ...friend
+                });
+            } else if (member.id === conn.peer) {
+                // insert the person we connected to, now that we know their info
+                this.#friends.set(conn.peer, {
+                    ...member,
+                    conn
+                });
+                this.onChangeFriends(this.#friends.size, 1);
+            } else {
+                // for new people, just go connect to them (& hopefully they aren't doing the same)
+                this.connect(member.id);
```

Comment:
So every message contains the full state data of the video and the group members + their metadata. If they're out of sync, the next message would correct them. And since we emit a message 500ms after joining, they shouldn't ever be out of sync for very long. If they are ... any member just needs to emit an event.

If this becomes a problem, we could swap this to be a heartbeat message from every member periodically.